File: 2021/day4.py
from util.runner import Runner
import numpy as np


class BingoCard:
    def __init__(self):
        self.rows = []
        self.card: type(np.ndarray) = np.ndarray([])

    def set_row(self, row):
        self.rows.append(row)

    def build_card(self):
        self.card = np.row_stack([np.array(r) for r in self.rows])

    def mark_card(self, number):
        if number in self.card:
            self.card[self.card == number] = -1

    def check_card(self):
        for row in self.card:
            if len([i for i in row if i >= 0]) == 0:
                return True
        for col in self.card.transpose():
            if len([i for i in col if i >= 0]) == 0:
                return True
        return False

    def get_score(self):
        # since we're using -1 to track removed elements
        # replace with 0 to avoid messing up the sum
        self.card[self.card == -1] = 0
        return self.card.sum()


class Day(Runner):
    def parse(self):
        cards = []
        buffer = None
        numbers = None
        for line in self.input:
            if numbers is None:
                numbers = [int(n) for n in line.split(',')]
                continue
            if line == "":
                if buffer is not None:
                    buffer.build_card()
                    cards.append(buffer)
                buffer = BingoCard()
            else:
                buffer.set_row([int(n) for n in line.split(' ') if n.strip() != ''])

        buffer.build_card()
        cards.append(buffer)

        return cards, numbers

    def part1(self):
        cards, numbers = self.parse()
        for n in numbers:
            for card in cards:
                card.mark_card(n)
                if card.check_card():
                    return n * card.get_score()

    def part2(self):
        cards, numbers = self.parse()
        for n in numbers:
            for card in cards:
                card.mark_card(n)
                if card.check_card():
                    cards.remove(card)
                if len(cards) == 1:
                    return n * card.get_score()


Day(4, lambda l: l.strip())

File: 2025/day4/day4.go
package main

import (
	"2025/runner"
	"2025/types"
	"2025/util"
	"strconv"
)

func part1(input string) string {
	rows := util.AsLines(input)

	total := 0

	for i, line := range rows {
		for j := range line {

			cell := string(rows[i][j])
			if cell == "@" && countAdjacents(rows, i, j, "@") < 4 {
				total += 1
			}
		}
	}

	return strconv.Itoa(total)
}

func part2(input string) string {
	rows := util.AsLines(input)

	prevTotal := -1
	total := 0
	var removed [][]int

	for total != prevTotal {
		prevTotal = total
		for i, line := range rows {
			for j := range line {

				cell := string(rows[i][j])
				if cell == "@" && countAdjacents(rows, i, j, "@") < 4 {
					total += 1
					removed = append(removed, []int{i, j})
				}
			}
		}

		for _, coords := range removed {
			i := coords[0]
			j := coords[1]
			rows[i] = replaceChar(rows[i], '.', j)
		}
	}

	return strconv.Itoa(total)
}

func replaceChar(s string, r rune, i int) string {
	out := []rune(s)
	out[i] = r
	return string(out)
}

func countAdjacents(grid []string, row int, col int, char string) int {
	total := 0
	for i := row - 1; i <= row+1; i++ {
		for j := col - 1; j <= col+1; j++ {
			if i < 0 || i >= len(grid) || j < 0 || j >= len(grid[0]) || (row == i && col == j) {
				continue
			}
			cell := string(grid[i][j])

			if cell == char {
				total++
			}
		}
	}
	return total
}

func main() {
	day := types.Day{
		Part1: part1,
		Part2: part2,
		Day:   "4",
	}

	runner.RunDay(day)
}
